### PR TITLE
Hoist default exported anonymous function declarations

### LINKED ISSUE #1784: Anonymous function declarations should be hoisted
https://rollupjs.org/repl?version=0.52.1&shareable=JTdCJTIybW9kdWxlcyUyMiUzQSU1QiU3QiUyMm5hbWUlMjIlM0ElMjJtYWluLmpzJTIyJTJDJTIyY29kZSUyMiUzQSUyMmltcG9ydCUyMGYlMjBmcm9tJTIwJTVDJTIyLiUyRmYuanMlNUMlMjIlM0IlNUNuaW1wb3J0JTIwZyUyMGZyb20lMjAlNUMlMjIuJTJGZy5qcyU1QyUyMiUzQiU1Q25jb25zb2xlLmxvZyhnKCkpJTNCJTIyJTdEJTJDJTdCJTIybmFtZSUyMiUzQSUyMmYuanMlMjIlMkMlMjJjb2RlJTIyJTNBJTIyaW1wb3J0JTIwZyUyMGZyb20lMjAlNUMlMjIuJTJGZy5qcyU1QyUyMiUzQiU1Q25leHBvcnQlMjBkZWZhdWx0JTIwZnVuY3Rpb24lMjAoZm4pJTIwJTdCJTVDbiUyMCUyMHJldHVybiUyMGZ1bmN0aW9uJTIwKHN0ciklMjAlN0IlMjByZXR1cm4lMjBzdHIlMjAlN0MlN0MlMjBnKGZuLm5hbWUpJTNCJTIwJTdEJTVDbiU3RCUyMiU3RCUyQyU3QiUyMm5hbWUlMjIlM0ElMjJnLmpzJTIyJTJDJTIyY29kZSUyMiUzQSUyMmltcG9ydCUyMGYlMjBmcm9tJTIwJTVDJTIyLiUyRmYuanMlNUMlMjIlM0IlNUNuZXhwb3J0JTIwZGVmYXVsdCUyMGYoZnVuY3Rpb24lMjBmbihzdHIpJTIwJTdCJTVDbiUyMCUyMHJldHVybiUyMHN0ciUyMCUyQiUyMHN0ciUzQiU1Q24lN0QpJTNCJTIyJTdEJTVEJTJDJTIyb3B0aW9ucyUyMiUzQSU3QiUyMmZvcm1hdCUyMiUzQSUyMmNqcyUyMiUyQyUyMmdsb2JhbHMlMjIlM0ElN0IlN0QlMkMlMjJuYW1lJTIyJTNBJTIybXlCdW5kbGUlMjIlMkMlMjJhbWQlMjIlM0ElN0IlMjJpZCUyMiUzQSUyMiUyMiU3RCU3RCUyQyUyMmV4YW1wbGUlMjIlM0FudWxsJTdE

That code throws `f` is not defined, because the function exported by `f.js` is not hoisted. If I give a name to that function or I remove `import f from "f.js"` from `main.js`, everything works.

I'm not 100% sure that this is a bug, since `@std/esm` works exactly like rollup. On the other hand, with `node --experimental-modules` it logs `"fn"`.

### PR DESCRIPTION
**Pull request by nicolo-ribaudo**
Fixes #1784

I didn't add any test because I already modified the expected output of some existing tests. If you want a new test, let me know.

### CODE REVIEW

File: src/ast/nodes/ExportDefaultDeclaration.js [hunk 1]
```diff
@@ -35,20 +35,22 @@ export default class ExportDefaultDeclaration extends Node {
 
 		// paren workaround: find first non-whitespace character position after `export default`
 		let declaration_start;
+		let id_index;
```
Comment (lukastaegert) [inline src/ast/nodes/ExportDefaultDeclaration.js]: "index" is a very vague name that is easily misinterpreted. How about `id_insertPos` or something similar that better describes what this variable is used for?

**Comment by nicolo-ribaudo** [comment]
Thank you for your feedback!
I already wanted to refactor that code because of https://github.com/rollup/rollup/blob/d0ba36b662d6f20fa207e2b1a6dbed334a3f1087/src/ast/nodes/ExportDefaultDeclaration.js#L31
I didn't because I preferred to make few changes in my first pr, but well... I can do it now :smile:

File: src/ast/nodes/ExportDefaultDeclaration.js [hunk 2]
```diff
@@ -28,60 +28,48 @@ export default class ExportDefaultDeclaration extends Node {
...
-				else {
-					if ( this.variable.getOriginalVariableName( es ) === name ) {
-						// prevent `var foo = foo`
-						code.remove( this.leadingCommentStart || this.start, this.next || this.end );
-						return; // don't render children. TODO this seems like a bit of a hack
-					} else {
-						code.overwrite( this.start, declaration_start, `${this.module.bundle.varOrConst} ${name} = ` );
-					}
+		if ( functionOrClassDeclaration.test(this.declaration.type) ) {
+			if ( treeshakeable ) {
```
Comment (nicolo-ribaudo) [inline src/ast/nodes/ExportDefaultDeclaration.js]: Does this word actually exist? :smile:

File: src/ast/nodes/ExportDefaultDeclaration.js [hunk 3]
```diff
@@ -28,60 +28,48 @@ export default class ExportDefaultDeclaration extends Node {
 		this.variable = this.scope.addExportDefaultDeclaration( this._declarationName || this.module.basename(), this );
 	}
 
-	// TODO this is total chaos, tidy it up
 	render ( code, es ) {
-		const treeshake = this.module.bundle.treeshake;
+		const remove = () => { code.remove( this.leadingCommentStart || this.start, this.next || this.end ); };
+		const removeExportDefault = () => { code.remove( this.start, declaration_start ); };
+		
+		const treeshakeable = this.module.bundle.treeshake && !this.included & !this.declaration.included;
```
Comment (lukastaegert) [inline src/ast/nodes/ExportDefaultDeclaration.js]: I guess it should be `&&` both times?

**Comment by lukastaegert** [comment]
BTW. I found my way into the project by basically refactoring large parts of the code base so I will certainly be the last person to keep others from making the code nicer 😉

File: src/ast/nodes/ExportDefaultDeclaration.js (hunk 3 above)
Comment (nicolo-ribaudo) [reply src/ast/nodes/ExportDefaultDeclaration.js]: Yes, I mistyped it. I will correct it later (3 hours)

**Comment by lukastaegert** [comment]
Looking at the example #1784 I think we should actually add the example given there as a functional test (i.e. in the `test/function` folder) as well as it is quite non-trivial and easily broken by future changes. I would change it slightly like this:
https://rollupjs.org/repl?version=0.52.1&shareable=JTdCJTIybW9kdWxlcyUyMiUzQSU1QiU3QiUyMm5hbWUlMjIlM0ElMjJtYWluLmpzJTIyJTJDJTIyY29kZSUyMiUzQSUyMmltcG9ydCUyMGYlMjBmcm9tJTIwJTVDJTIyLiUyRmYuanMlNUMlMjIlM0IlNUNuaW1wb3J0JTIwZyUyMGZyb20lMjAlNUMlMjIuJTJGZy5qcyU1QyUyMiUzQiU1Q24lNUNuYXNzZXJ0LmVxdWFsKGcoKSUyQyUyMCdmbicpJTNCJTIyJTdEJTJDJTdCJTIybmFtZSUyMiUzQSUyMmYuanMlMjIlMkMlMjJjb2RlJTIyJTNBJTIyaW1wb3J0JTIwZyUyMGZyb20lMjAlNUMlMjIuJTJGZy5qcyU1QyUyMiUzQiU1Q25leHBvcnQlMjBkZWZhdWx0JTIwZnVuY3Rpb24lMjAoZm4pJTIwJTdCJTVDbiUyMCUyMHJldHVybiUyMGZ1bmN0aW9uJTIwKHN0ciklMjAlN0IlMjByZXR1cm4lMjBzdHIlMjAlN0MlN0MlMjBnKGZuLm5hbWUpJTNCJTIwJTdEJTVDbiU3RCUyMiU3RCUyQyU3QiUyMm5hbWUlMjIlM0ElMjJnLmpzJTIyJTJDJTIyY29kZSUyMiUzQSUyMmltcG9ydCUyMGYlMjBmcm9tJTIwJTVDJTIyLiUyRmYuanMlNUMlMjIlM0IlNUNuZXhwb3J0JTIwZGVmYXVsdCUyMGYoZnVuY3Rpb24lMjBmbihzdHIpJTIwJTdCJTVDbiUyMCUyMHJldHVybiUyMHN0ciUyMCUyQiUyMHN0ciUzQiU1Q24lN0QpJTNCJTIyJTdEJTVEJTJDJTIyb3B0aW9ucyUyMiUzQSU3QiUyMmZvcm1hdCUyMiUzQSUyMmVzJTIyJTJDJTIyZ2xvYmFscyUyMiUzQSU3QiU3RCUyQyUyMm5hbWUlMjIlM0ElMjJteUJ1bmRsZSUyMiUyQyUyMmFtZCUyMiUzQSU3QiUyMmlkJTIyJTNBJTIyJTIyJTdEJTdEJTJDJTIyZXhhbXBsZSUyMiUzQW51bGwlN0Q=

What do you think? I plan on doing a release some time tomorrow, I can also easily add this myself then. You already did rollup a great service with this PR!

**Comment by nicolo-ribaudo** [comment]
Test added!

File: src/ast/nodes/ExportDefaultDeclaration.js (hunk 2 above)
Comment (Andarist) [reply src/ast/nodes/ExportDefaultDeclaration.js]: I've used it quite a bit on multiple occasions 😉

**Pull request merged**